crates/containerd-shim-wasmedge/src/instance.rs

```diff
@@ -103,171 +131,103 @@ pub fn reset_stdio() {
-                let _ret = match vm.run_func(Some("main"), "_start", params!()) {
-                    Ok(_) => std::process::exit(0),
-                    Err(_) => std::process::exit(137),
-                };
-            }
-        }
+
+        let syscall = create_syscall();
+        let mut container = ContainerBuilder::new(self.id.clone(), syscall.as_ref())
+            .with_executor(vec![Box::new(WasmEdgeExecutor {
+                stdin: maybe_open_stdio(self.stdin.as_str()).context("could not open stdin")?,
+                stdout: maybe_open_stdio(self.stdout.as_str()).context("could not open stdout")?,
+                stderr: maybe_open_stdio(self.stderr.as_str()).context("could not open stderr")?,
+            })])?
+            .with_root_path(self.rootdir.as_str())?
+            .as_init(&self.bundle)
+            .with_systemd(false)
+            .build()?;
+
+        let code = self.exit_code.clone();
+        let pid = container.pid().unwrap();
+        container.start()?;
+
+        thread::spawn(move || {
+            let (lock, cvar) = &*code;
+
+            let status = match waitid(WaitID::Pid(pid), WaitPidFlag::WEXITED) {
+                Ok(WaitStatus::Exited(_, status)) => status,
+                Ok(WaitStatus::Signaled(_, sig, _)) => sig as i32,
+                Ok(_) => 0,
+                Err(e) => {
+                    if e == Errno::ECHILD {
+                        0
+                    } else {
+                        panic!("waitpid failed: {}", e);
+                    }
+                }
+            } as u32;
+            let mut ec = lock.lock().unwrap();
+            *ec = Some((status, Utc::now()));
+            drop(ec);
+            cvar.notify_all();
+        });
+
+        Ok(0)
     }
 
     fn kill(&self, signal: u32) -> Result<(), Error> {
         if signal as i32 != SIGKILL && signal as i32 != SIGINT {
-            println!("{:?}", signal);
             return Err(Error::InvalidArgument(
                 "only SIGKILL and SIGINT are supported".to_string(),
             ));
         }
 
-        let lr = self.pidfd.lock().unwrap();
-        let fd = lr
-            .as_ref()
-            .ok_or_else(|| Error::FailedPrecondition("module is not running".to_string()))?;
-        fd.kill(SIGKILL as i32)
+        let mut container = load_container(self.rootdir.as_str(), self.id.as_str())?;
+        match container.kill(Signal::try_from(signal as i32)?, true) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                if container.status() == ContainerStatus::Stopped {
+                    return Err(Error::Others("container not running".into()));
+                }
+                Err(Error::Others(e.to_string()))
+            }
+        }
     }
 
     fn delete(&self) -> Result<(), Error> {
-        let spec = match load_spec(self.bundle.clone()) {
-            Ok(spec) => spec,
+        match container_exists(self.rootdir.as_str(), self.id.as_str()) {
```

Comment:
If a container is present to meet OCI specifications, it must return Ok.

Reply:
Sure, I was just thinking that they are called one after another, so this could maybe be a place for optimization. But this is a very minor thing and I'm fine either way.